libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs

```diff
@@ -0,0 +1,222 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Dialogs
+{
+    public class DialogContext
+    {
+        private object _finalResult;
+
+        public DialogSet Dialogs { get; set; }
+        public ITurnContext Context { get; set; }
+        public Stack<DialogInstance> Stack { get; set; }
+
+        /// <summary>
+        /// Creates a new DialogContext instance.
+        /// </summary>
+        /// <param name="dialogs">Parent dialog set.</param>
+        /// <param name="context">Context for the current turn of conversation with the user.</param>
+        /// <param name="stack">Current dialog stack.</param>
+        public DialogContext(DialogSet dialogs, ITurnContext context, Stack<DialogInstance> stack)
+        {
+            Dialogs = dialogs;
+            Context = context;
+            Stack = stack;
+        }
+
+        /// <summary>
+        /// Returns the cached instance of the active dialog on the top of the stack or `undefined` if the stack is empty.
+        /// </summary>
+        public DialogInstance Instance
+        {
+            get
+            {
+                if (Stack.Any())
+                {
+                    return Stack.Peek();
+                }
+                return null;
+            }
+        }
+
+        public DialogResult DialogResult
+        {
+            get
+            {
+                return new DialogResult
+                {
+                    Active = Stack.Count > 0,
+                    Result = _finalResult
+                };
+            }
+        }
+
+
+        /// <summary>
+        /// Pushes a new dialog onto the dialog stack.
+        /// </summary>
+        /// <param name="dialogId">ID of the dialog to start.</param>
+        /// <param name="dialogArgs">(Optional) additional argument(s) to pass to the dialog being started.</param>
+        public async Task Begin(string dialogId, object dialogArgs = null)
+        {
+            // Lookup dialog
+            var dialog = Dialogs.Find(dialogId);
+            if (dialog == null)
+            {
+                throw new Exception($"DialogContext.begin(): A dialog with an id of '{dialogId}' wasn't found.");
+            }
+            
+            // Push new instance onto stack. 
+            var instance = new DialogInstance
+            {
+                Id = dialogId,
+                State = new object()
```

Comment:
In... what scenario is this actually useful? This code seem actively harmful.

Reply:
This is another area where I had questions for the C# based implementation.  It's not only useful it's essential for waterfalls to work.  You need a scratch memory space to store state while you're progressing through a waterfall. For instance you need to be able to store the form you're building up as you ask a user for individual field values.  

In JS this is just a JSON object { } that they can write values to and its sand boxed on a per/dialog bases to avoid getting accidentally overwritten by other dialogs on the stack.  For C# I think a dictionary makes more sense but it has to be serialized as part of the stack.